### DISCUSSION TITLE
Western Digital FD1797 Floppy Disk controller

### DISCUSSION THREAD

**Discussion by steffenspies** [keirf/greaseweazle#227] [Q&A]
Hello around,
CP/M trouble... what else.

I try to read my old CP/M disks but fail massively and I've no idea what else can be done.
Maybe someone has a good hint where to start.

What I know: 
5 1/4" [Teac FD-55B](https://minuszerodegrees.net/manuals/TEAC/TEAC%20FD-55%20Series.pdf) is used from the original system (which still works, so I know the disk is working).
Floppy Disk Controller is [Western Digital FD1797](https://www.abc80.net/archive/luxor/diskdrives/drives/western-digital-FD1791.pdf) 
A formated disk has 320kB free, is double sided, 40 Tracks/side, 48tpi, MFM, 300 RPM
The WD document mentions a "IBM System 34 Double Density (MFM)" format.

I tried almost any possible settings, different drives, diskettes and had the chance to try a kryflux too beside my Greaseweazlke V4  with no luck at all.

Track 00 can't be found seems to me the biggest problem.

Thanks for time and brain crunching!
Steffen

**Comment by keirf** [comment]
What type of greaseweazle? Tried another 5.25 drive?

**Reply by steffenspies** [reply]
Just updated my posting.
Greaseweazle V4 and different drives. 2x TEAC FD-55B and one FD-55GFR

**Reply by keirf** [reply]
Have you separately powered the drive so it gets sufficient current and both 12v and 5v?

Is the ribbon straight or does it have a twist in it? Have you played with the drive select option in gw?

**Reply by steffenspies** [reply]
It is pwered externaly and both voltages (5V and +12V) are strong and stable.
I tried both two twisted and straight cables (drive b). Cables are tested.
All works with standard MS-DOS formated disks.

**Reply by keirf** [reply]
Dump to scp format and upload so we can take a look. Hxc visualiser tool would be interesting. Probably its a track format not in gw's static list of formats.

**Reply by steffenspies** [reply]
I've done some runs and hope this all makes sense.
HxC hex editor shows some true datas, but the visualization does look strange to me.
[SCP dump](https://github.com/steffenspies/CP-M-Disk-Analyse/blob/main/cpm06.zip)

[HxC visualizer ](https://github.com/steffenspies/CP-M-Disk-Analyse/blob/main/Screenshot_20221025_211053.png)

**Reply by keirf** [reply]
It's probably an rpm thing actually. Could try --adjust-speed 300 on the command line

**Reply by steffenspies** [reply]
My bad - I'm sorry, missed that, should have known better.


[HxC](https://github.com/steffenspies/CP-M-Disk-Analyse/blob/main/cpm08.jpg)
[SCP](https://github.com/steffenspies/CP-M-Disk-Analyse/blob/main/cpm08.scp)

**Reply by keirf** [reply]
5 1kb sectors per track probably. Not a format in gw's list and it doesn't have a "dump what you find to img" mode. You could export this as img using hxc though. Or use my disk-utilities project (vastly less user friendly!)

**Reply by keirf** [reply]
Saying that, two formats in gw that it might be: `akai.800` or `acorn.adfs.800`

Eg `gw read --tracks=c=0-39 --format=akai.800 a.img`

**Reply by keirf** [reply]
Or no need to dump again, convert the existing scp image file:
`gw convert --tracks=c=0-39 --format=akai.800 my.scp my.img`

**Reply by steffenspies** [reply]
Thank you very much.
Found [this](https://github.com/steffenspies/CP-M-Disk-Analyse/blob/main/IBM_Track_Format-FD1797.jpg) in the doc from the [Western Digital FD1797](https://www.abc80.net/archive/luxor/diskdrives/drives/western-digital-FD1791.pdf) page 17
Would this make sense?
Any chance to make it part of the project since it was a common format in the early days?

**Reply by keirf** [reply]
IBM System 34 format is standard flexible IBM track format. Sectors sizes and identities are defined on track via sector headers. I should add gw format specifiers which accept IBM tracks "as they come" but right now the formats are more prescriptive: "I want 9 512 byte sectors" and so on.

**Reply by steffenspies** [reply]
Thank you anyway for this cool project!

**Reply by keirf** [reply]
Does the gw convert command line above work for you?

**Reply by steffenspies** [reply]
`gw read --tracks=c=0-39 --format=akai.800 a.img`
`gw convert --tracks=c=0-39 --format=akai.800 my.scp my.img`
Seems to work, no errors, but I can't open the image file.

`Cyl-> 0         1         2         3
H. S: 0123456789012345678901234567890123456789
0. 0: ........................................
0. 1: ........................................
0. 2: ........................................
0. 3: ........................................
0. 4: ........................................
1. 0: ........................................
1. 1: ........................................
1. 2: ........................................
1. 3: ........................................
1. 4: ........................................
Found 400 sectors of 400 (100%)`

**Reply by keirf** [reply]
Tried it, and it does work. So the format you need to use is `akai.800`

**Reply by keirf** [reply]
How do you open the img file? I'm pretty confident it's a good sector dump. If you hexdump the img file you can see the file contents and so on.

**Reply by keirf** [reply]
The thing is: It will be a CP/M filesystem. Not FAT. So Windows and MSDOS won't know what to do with it. You need tools that understand CP/M filesystems, or you need a suitable CP/M emulator.

**Reply by steffenspies** [reply]
I use WinImage

**Reply by steffenspies** [reply]
OK, ttoo much dumb stuff now... time to go to bed... sure... I need a CP/M system... I will try it tomorrow with my IMSAI clone.

**Reply by keirf** [reply]
Good night. I'm pretty sure cp/m systems had a fairly standardised filesystem so I'm confident you'll be able to find tools to extract your files. Let us know what you discover!

**Reply by steffenspies** [reply]
That was a tough night - thinking about this all night long.

Tried to mount the disk with my CP/M tools, but that didn't work with bad format.

The generated disk image from the .scp file is 400kB in size, but the disks have 320kB formated / 360kB raw.
The large sector size confuses me and I don't know if this is a realistic value - I don't have any experiences here (yet).
But the hex editor shows some very useful data, which is a good sign.
I think my next steps should be to crawl into the old computer and find out what the parameters are. Luckily it is still alive and kicking.

If anyone has other ideas, please let me know!

p.s.: To check if my tool chain works correct, I tried an old MS-DOS disc which worked well.

**Reply by keirf** [reply]
From where do you get the 360kB raw figure?

**Reply by keirf** [reply]
Most of the disk is empty so you could truncate the IMG to 360kB if you think it might help the image be recognised. Actually only the first 126kB of the disk image contains any data at all.

**Reply by keirf** [reply]
Okay, mystery solved. You can decode your files using cpmtools and the format is `ampdsdd` (also `pmc101`, `amp3` and `amp5` all work too):
```
cpmls -f ampdsdd mydisk.img
cpmcp -f ampdsdd mydisk.img 0:* mydisk/
```

**Reply by steffenspies** [reply]
Thank you, that was very kind and very helpful.
Now that I have a solution and I'm very happy that there is way now to read all my old Floppies, I still want to find out why my informations are wrong (360kB, sector numbers, sector size). Since my CP/M computer is a homebrew project based on a publication in a German magazine, some informations seem to come from  wrong sources.

And I'm sorry that you had to spend so much time for my problem. Need to dive deeper into this. Thank you again, @keirf

**Reply by steffenspies** [reply]
OK, what a mess...
I found out, that my old documentation simply was wrong. Had have a look into the format tool I used back in the days and now all you, @keirf found out was correct. An formated disk has 390kB free and 1024kB/sector is also correct(obvisly).
Meanwhile all 59 disks are read, stored and are now converted to use with the IMSAI clone. Found only very few bad sectors, which I find amazing after so many years.
